### DISCUSSION TITLE
【打卡】阅读

### DISCUSSION THREAD

**Discussion by swustzzh** [swustzzh/coder_books#19] [General]
书名：图解HTTP

章节： 第一章至第五章

总结：

- TCP/IP的分层管理：
  - 应用层：决定了向用户提供应用服务时通信的活动。预存了各类通用的应用服务，如FTP、DNS、HTTP
  - 传输层：对上层应用层提供处于网络连接中两台计算机之间的数据传输。存在TCP、UDP两种协议
  - 网络层：处理网络上流动的数据包。
  - 数据链路层：用来处理链接网络的硬件部分。
- HTTP无状态协议，Cookie的设置与使用
- 持久化链接，HTTP1.1默认开启keep-alive，目的在于一次TCP/IP连接的过程中可以发送多个http请求
- HTTP状态码
  - 1xx: 信息性状态码，接受的请求正在处理
  - 2xx: 成功状态码 请求正常处理完毕
  - 3xx: 重定向状态码 需要进行附加操作已完成请求
  - 4xx: 客户端错误状态码 服务器无法处理请求
  - 5xx: 服务器错误状态码 服务器处理请求出错

思考：

- 本书的知识点都较为基础，计算机专业的基本上等于复习网络相关的知识。

**Comment by swustzzh** [comment]
章节：第六章 HTTP头部  
总结：
 - HTTP首部字段：构成HTTP报文的要素之一。在客户端和服务器之间一HTTP协议进行通信的过程中，无论是请求还是相应都会使用首部字段，能起到传递额外重要信息的作用。
 - HTTP首部字段重复？不同浏览器处理逻辑不同，有些取第一次，有些最最后一次（所以还是别设置重复的字段了，可以同个字段多个值）

**Reply by swlws** [reply]
HTTP2 的多路复用，“多路”、“复用”分别什么意思？

“多路”，在同一时间，多个数据流（即HTTP请求和响应）可以在单个TCP连接上并行传输
“复用”，同一TCP连接上，多个数据流可以同时共享连接资源

来自 ChatGPT 的回答，感觉是废话

**Reply by swustzzh** [reply]
HTTP/2中的多路复用是一种允许多个请求和响应在同一时间通过单个TCP连接并行传输的机制。这是通过将HTTP消息分割为独立的帧，然后在同一连接上交错发送，最后在另一端重新组装的方式实现的。

这种机制有以下几个主要优点：

并行性：在HTTP/1.x中，如果要并行发送多个请求，需要使用多个TCP连接，这会消耗更多的资源。而在HTTP/2中，只需要一个TCP连接，就可以并行发送多个请求或响应。

资源优化：因为只需要一个TCP连接，所以可以减少TCP握手的次数，降低延迟，同时还可以更好地利用网络资源。

流控制：HTTP/2还引入了流控制的概念，每个流都有一个关联的优先级和权重，这使得服务器可以根据资源的可用情况和流的优先级来决定响应的顺序。

减少阻塞：在HTTP/1.x中，请求和响应的发送是按顺序进行的，如果前面的请求因为某些原因被阻塞，那么后面的请求也会被阻塞，这被称为“队头阻塞”。而在HTTP/2中，由于引入了帧的概念，即使某个帧被阻塞，也不会影响其他帧的传输。

总的来说，HTTP/2中的多路复用技术，提高了Web应用的性能，降低了延迟，同时也使得网络资源的利用更加高效。

--chatGPT。核心的点是分帧。

**Comment by swustzzh** [comment]
章节：第七章 确保Web安全的HTTPS  
总结：  
- HTTP的缺点：
  - 明文传输
  - 无状态（token是服务器端和客户端验证的方式，和HTTP没关系）
  - 无法验证报文的完整性
- 使用SSL或TSL进行加密

**Reply by swustzzh** [reply]
浑身疼，淦！

**Comment by swustzzh** [comment]
章节：第八章  - 第十一章
总结：  
- 后续几个章节干货不多，且本书主要是HTTP1.1的内容，可以额外了解一下HTTP2.0/3.0相关的内容
常见前端攻击方式
- Cookie劫持：给cookie添加http-only属性(面试被问到过)、cookie中添加校验信息ip,user-agent,refer等、cookie中的session-id定时更换
- XSS跨站脚本攻击：标签过滤、编码转义、纯前端渲染、避免不可信的数据拼接到字符串中传递给这些API
- CSRF：验证码、cookie设置samesite、使用post请求、检查http请求头中的refer、token、自定义属性并验证
- 点击挟持：js禁止内嵌、设置http响应头x-frame-options


没有绝对的安全，常见的安全措施是为了增加破解的成本，过滤低成本的攻击。
